### PR TITLE
feat: new media upload api

### PR DESCRIPTION
**Pull request by shyim**
### 1. Why is this change necessary?

The current media upload API (`/api/_action/media/{mediaId}/upload`) requires creating a media entity first before uploading a file. This creates an unnecessary two-step process and doesn't provide modern upload capabilities like deduplication, direct URL uploads, or external linking. A new streamlined API is needed to simplify media management workflows.

### 2. What does this change do, exactly?

This PR introduces a new Media Upload API v2 with three new endpoints that provide more flexible and efficient media upload capabilities:

- **`/api/_action/media/upload`** - Direct file upload that creates media entity and uploads file in one step
- **`/api/_action/media/upload_by_url`** - Download and upload file from external URL
- **`/api/_action/media/external-link`** - Create media entity that links to external URL without downloading

**Key Features:**
- ✅ One-step upload process (no need to create media entity first)
- ✅ Built-in deduplication based on file hash or URL path
- ✅ Support for custom IDs, media folders, and privacy settings
- ✅ Comprehensive parameter validation and error handling
- ✅ Full unit test coverage (43 tests, 178 assertions)
- ✅ OpenAPI 3.1 documentation included

**Components Added:**
- `MediaUploadV2Controller` - New controller with three upload methods
- `MediaUploadService` - Service layer handling upload logic
- `MediaUploadParameters` - DTO for upload configuration
- Comprehensive unit tests for all components
- Updated OpenAPI schema documentation

### 3. Describe each step to reproduce the issue or behaviour.

**Current workflow (old API):**
```bash
# Step 1: Create empty media entity
POST /api/media
{
  "id": "01234567890123456789012345678901"
}

# Step 2: Upload file to existing entity
POST /api/_action/media/01234567890123456789012345678901/upload
Content-Type: multipart/form-data
file: [binary data]
```

**New workflow (this PR):**
```bash
# Single step: Create entity and upload file
POST /api/_action/media/upload
Content-Type: multipart/form-data
file: [binary data]
```

### 1. Direct File Upload
Upload a file and create media entity in one request:

```bash
curl -X POST \
  -H "Authorization: Bearer {access_token}" \
  -F "file=@image.jpg" \
  -F "fileName=my-image.jpg" \
  -F "private=false" \
  -F "mediaFolderId=01234567890123456789012345678901" \
  -F "deduplicate=true" \
  "http://localhost/api/_action/media/upload"
```

**Response:**
```json
{
  "id": "01234567890123456789012345678901"
}
```

### 2. Upload from URL
Download and upload file from external URL:

```bash
curl -X POST \
  -H "Authorization: Bearer {access_token}" \
  -H "Content-Type: application/json" \
  -d '{
    "url": "https://example.com/image.jpg",
    "fileName": "downloaded-image.jpg",
    "mediaFolderId": "01234567890123456789012345678901",
    "deduplicate": true
  }' \
  "http://localhost/api/_action/media/upload_by_url"
```

**Response:**
```json
{
  "id": "01234567890123456789012345678901"
}
```

### 3. External Link (CDN/External Storage)
Create media entity that links to external URL:

```bash
curl -X POST \
  -H "Authorization: Bearer {access_token}" \
  -H "Content-Type: application/json" \
  -d '{
    "url": "https://cdn.example.com/image.jpg",
    "fileName": "external-image.jpg",
    "mimeType": "image/jpeg",
    "private": false,
    "deduplicate": true
  }' \
  "http://localhost/api/_action/media/external-link"
```

**Response:**
```json
{
  "id": "01234567890123456789012345678901"
}
```

### 4. Advanced Parameters
All endpoints support these optional parameters:

```json
{
  "id": "custom-id-here",                    // Custom media entity ID
  "fileName": "custom-filename.jpg",         // Override filename
  "private": true,                           // Private media (default: false)
  "mediaFolderId": "folder-id",              // Organize in media folder
  "mimeType": "image/jpeg",                  // Override detected MIME type
  "deduplicate": true                        // Skip if duplicate exists
}
```

### 5. JavaScript/TypeScript Usage
```typescript
// Direct file upload
const formData = new FormData();
formData.append('file', fileInput.files[0]);
formData.append('mediaFolderId', 'folder-id');
formData.append('deduplicate', 'true');

const response = await fetch('/api/_action/media/upload', {
  method: 'POST',
  headers: { 'Authorization': `Bearer ${token}` },
  body: formData
});

const { id } = await response.json();

// URL upload
const urlResponse = await fetch('/api/_action/media/upload_by_url', {
  method: 'POST',
  headers: {
    'Authorization': `Bearer ${token}`,
    'Content-Type': 'application/json'
  },
  body: JSON.stringify({
    url: 'https://example.com/image.jpg',
    deduplicate: true
  })
});
```

### 4. Please link to the relevant issues (if any).

This implements a new media upload API to modernize and streamline media management workflows. No specific issue linked, this is a feature enhancement.

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfilled them

**Test Coverage:**
- ✅ 43 unit tests covering all components
- ✅ 178 assertions validating functionality
- ✅ PHPStan level 9 compliance
- ✅ All edge cases and error scenarios tested

**Documentation:**
- ✅ OpenAPI 3.1 schema definitions added
- ✅ Comprehensive parameter documentation
- ✅ Usage examples for all endpoints
- ✅ Error response documentation

### CODE REVIEW

**Comment by OliverSkroblin** [comment]
Hey! 👍 This looks great - the current media API flow is definitely painful.

One quick point, batch API: Would love to see this in the /api/_action/sync endpoint too. For bulk imports, I often need the media IDs immediately for linking to products in the same batch operation.

And maybe also more sources?
```
// Base64 encoded data
$mediaId = $this->mediaUploadService->uploadFromBase64($base64Data, $context, new MediaUploadParameters(fileName: 'image.jpg'));

// Raw binary data
$mediaId = $this->mediaUploadService->uploadFromBinary($binaryData, $context, new MediaUploadParameters(fileName: 'image.jpg'));

// Stream resource
$mediaId = $this->mediaUploadService->uploadFromStream($stream, $context, new MediaUploadParameters(fileName: 'image.jpg'));
```

Otherwise solid improvement - will make media handling much cleaner!

File: src/Core/Content/Media/MediaException.php [hunk 1]
```diff
@@ -436,4 +436,22 @@ public static function unknownLocationType(): self
             'Unknown location type'
         );
     }
+
+    public static function fileNotProvided(): self
+    {
+        return new self(
+            Response::HTTP_BAD_REQUEST,
+            'CONTENT__MEDIA_NO_FORM_DATA_FIELD_PROVIDED',
+            'No form data field "field" provided.'
```
Comment (MartinKrzykawski) [inline src/Core/Content/Media/MediaException.php]: The error description could be improved - this does not reflect the issue

File: src/Core/Content/Media/Api/MediaUploadV2Controller.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        if (!\is_string($url)) {
+            throw MediaException::invalidUrl($url ?? '');
+        }
+
+        return new JsonResponse([
+            'id' => $this->mediaUploadService->linkURL($url, $context, $this->buildMediaUploadParamsFromRequest($request)),
+        ]);
+    }
+
+    private function buildMediaUploadParamsFromRequest(Request $request): MediaUploadParameters
```
Comment (MartinKrzykawski) [inline src/Core/Content/Media/Api/MediaUploadV2Controller.php]: IMO, it would be cleaner to move this logic into the MediaUploadParameters class as a fromRequest() method, since it’s the responsibility of MediaUploadParameters to know how to populate its own fields—not the controller

File: src/Core/Content/Media/Upload/MediaUploadService.php [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        $tempFile = tempnam(sys_get_temp_dir(), '');
+
+        if (!$tempFile) {
+            throw MediaException::cannotCreateTempFile();
+        }
+
+        $params->fillDefaultFileName(basename($url));
+
+        try {
+            $media = $this->fileFetcher->fetchFileFromURL(new Request([], ['url' => $url]), $tempFile);
```
Comment (MartinKrzykawski) [inline src/Core/Content/Media/Upload/MediaUploadService.php]: This seems hacky, as a new Request instance is created and passed just to satisfy the method signature. Wouldn’t it be cleaner to have a separate method that directly accepts a string URL instead?

File: src/Core/Content/Media/Api/MediaUploadV2Controller.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Attribute\Route;
+
+/**
+ * @internal
+ */
+#[Route(defaults: ['_routeScope' => ['api']])]
+#[Package('discovery')]
+readonly class MediaUploadV2Controller
```
Comment (MartinKrzykawski) [inline src/Core/Content/Media/Api/MediaUploadV2Controller.php:20]: A changelog would be great to inform external devs that this is now possible :)

**Review by MartinKrzykawski — COMMENTED** [review]
Heyy, really appreciate the effort! 👍 Looks good overall — I just added a couple of minor notes. Great work!

**Review by dgrothaus-sw — APPROVED** [review]

File: src/Core/Content/Media/File/FileFetcher.php [hunk 5]
```diff
@@ -77,20 +71,30 @@ public function fetchFileFromURL(Request $request, string $fileName): MediaFile
...
             $writtenBytes,
             // Change length of db field `media`.`file_hash` if algorithm is changed
             Hasher::hashFile($fileName, 'md5')
         );
     }
 
+    public function fetchFileFromURL(Request $request, string $fileName): MediaFile
+    {
+        if (!$this->enableUrlUploadFeature) {
+            throw MediaException::disableUrlUploadFeature();
```
Comment (mitelg) [inline src/Core/Content/Media/File/FileFetcher.php]: why don't you check this in the new method?

File: src/Core/Content/Media/Api/MediaUploadV2Controller.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+{
+    public function __construct(private MediaUploadService $mediaUploadService)
+    {
+    }
+
+    #[Route(path: '/api/_action/media/upload', name: 'api.action.media.upload_v2', methods: ['POST'])]
+    public function upload(Request $request, Context $context): Response
```
Comment (mitelg) [inline src/Core/Content/Media/Api/MediaUploadV2Controller.php]:
```suggestion
    public function upload(Request $request, Context $context): JsonResponse
```

File: src/Core/Content/Media/Api/MediaUploadV2Controller.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        return new JsonResponse(['id' => $this->mediaUploadService->uploadFromRequest($request, $context, MediaUploadParameters::fromRequest($request))]);
+    }
+
+    #[Route(path: '/api/_action/media/upload_by_url', name: 'api.action.media.upload_v2_url', methods: ['POST'])]
+    public function uploadUrl(Request $request, Context $context): Response
```
Comment (mitelg) [inline src/Core/Content/Media/Api/MediaUploadV2Controller.php]:
```suggestion
    public function uploadUrl(Request $request, Context $context): JsonResponse
```

File: src/Core/Content/Media/Api/MediaUploadV2Controller.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+        if (!\is_string($url)) {
+            throw MediaException::invalidUrl($url ?? '');
+        }
+
+        return new JsonResponse(['id' => $this->mediaUploadService->uploadFromURL($url, $context, MediaUploadParameters::fromRequest($request))]);
+    }
+
+    #[Route(path: '/api/_action/media/external-link', name: 'api.action.media.external-link', methods: ['POST'])]
+    public function externalLink(Request $request, Context $context): Response
```
Comment (mitelg) [inline src/Core/Content/Media/Api/MediaUploadV2Controller.php]:
```suggestion
    public function externalLink(Request $request, Context $context): JsonResponse
```

File: changelog/2025-06-23-media-upload-api-v2.md [hunk 9]
```diff
@@ -0,0 +1,10 @@
+---
+title: New Media Upload API v2
+---
+# Core
```
Comment (mitelg) [inline changelog/2025-06-23-media-upload-api-v2.md]:
```suggestion
# API
```

under `Core` the new service could be mentioned. also a `Upgrade Information` section should be added to inform in the upgrade guide about the new feature

File: src/Core/Content/Media/Upload/MediaUploadParameters.php [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        public ?string $id = null,
+        public ?string $mediaFolderId = null,
+        public ?bool $private = null,
+        public ?string $fileName = null,
+        public ?string $mimeType = null,
+        public ?bool $deduplicate = null
+    ) {
+    }
+
+    public static function fromRequest(Request $request): self
```
Comment (mitelg) [inline src/Core/Content/Media/Upload/MediaUploadParameters.php]: isn't it possible to use this Symfony feature? https://symfony.com/blog/new-in-symfony-6-3-mapping-request-data-to-typed-objects

File: src/Core/Content/Media/Upload/MediaUploadService.php [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+
+        return $this->upload($media, $context, $params);
+    }
+
+    /**
+     * Upload a new media file provided as form-data in the Request object
+     */
+    public function uploadFromRequest(Request $request, Context $context, MediaUploadParameters $params = new MediaUploadParameters()): string
+    {
+        /** @var UploadedFile|null $file */
```
Comment (mitelg) [inline src/Core/Content/Media/Upload/MediaUploadService.php]: please do not use var annotations. 

instead you could adjust the below if condition and check for the type you need

File: tests/unit/Core/Content/Media/Upload/MediaUploadServiceTest.php [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+    }
+
+    public function testUploadFromLocalPath(): void
+    {
+        $filePath = __DIR__ . '/fixtures/test-image.jpg';
+        $mediaId = Uuid::randomHex();
+        $params = new MediaUploadParameters();
+
+        // Create test file
+        file_put_contents($filePath, 'test content');
```
Comment (mitelg) [inline tests/unit/Core/Content/Media/Upload/MediaUploadServiceTest.php]: please use the Symfony FileSystem instead of native functions.

**Review by mitelg — CHANGES_REQUESTED** [review]

File: changelog/2025-06-23-media-upload-api-v2.md (hunk 9 above)
Comment (mitelg) [reply changelog/2025-06-23-media-upload-api-v2.md]: what about the upgrade information? :blush:

Comment (shyim) [reply changelog/2025-06-23-media-upload-api-v2.md]: why? you don't have to do anything when you upgrade. everything works as before

Comment (mitelg) [reply changelog/2025-06-23-media-upload-api-v2.md]: don't we want to inform the people about the new feature? isn't that also the purpose of this document?

//cc @alvarothomas @keulinho

**Review by MartinKrzykawski — APPROVED** [review]

**Review by mitelg — APPROVED** [review]

File: changelog/2025-06-23-media-upload-api-v2.md (hunk 9 above)
Comment (shyim) [reply changelog/2025-06-23-media-upload-api-v2.md]: We wrote in DM it's fine, I will mention it in the next release notes
